src/passport-saml/saml.ts

```diff
@@ -877,9 +903,9 @@ class SAML {
 
       return this.certsToCheck()
         .then(certs => {
-          const hasValidQuerySignature = certs.some(cert => {
+          const hasValidQuerySignature = certs!.some(cert => {
             return this.validateSignatureForRedirect(
-              urlString, container.Signature, container.SigAlg, cert
+              urlString, container.Signature as string, container.SigAlg as string, cert
```

Comment:
These things feel more like a hope than an assurance. Bad incoming data would cause this not to be true.

Overall, I'm looking at adding these types, not as a way to make TypeScript happy, but to actually express the truth. I'd rather leave gaps in the types where we don't know for sure, than to express something that may not be true just because it should be.

I won't comment on any more of these `as type` at this time.

Reply:
express has the support of parsing multiple Signature parameters. Do we need to write code to cover every case like this? It will be tens of Array.isArray() conditions.

Reply:
No, that would be the wrong approach. Really, what we want is to not assume that we have a string, but tell the user when we don't have a string, so all we need is this:

```javascript
if (typeof container.Signature === 'string' && typeof container.SigAlg === 'string') {
    return this.validateSignatureForRedirect(
        urlString, container.Signature, container.SigAlg, cert
    );
} else {
    throw new Error("Unexpected Signature or Signature Algorithm format.")
}
```

That really increases the quality of the code and helps the user. Simply adding `as string` hurts the developer and doesn't help the user.